### ISSUE TITLE
[Misc]: setting environment variables in multi-node serving

### ISSUE THREAD

**Issue by youkaichao** [vllm-project/vllm#6803]
### Anything you want to discuss about vllm.

As we embrace large models like Llama 3.1 405B, lots of users are trying multi-node inference now.

Compared with single-node inference, multi-node inference is much more difficult to set up, due to the nature of complicated machine configuration.

The [documentation](https://docs.vllm.ai/en/stable/serving/distributed_serving.html#multi-node-inference-and-serving) serves as a starting point. And we have a [debugging guide](https://docs.vllm.ai/en/stable/getting_started/debugging.html) with a sanity check script for testing the configuration. Even with this help, users might still find it difficult to set up the cluster, especially w.r.t. network configuration to make the machines talk to each other.

This discussion issue tries to clarify one aspect: how to set environment variables in multi-node setting, and how does environment variable inheritance work  in multi-node serving.

> NOTE: https://github.com/vllm-project/vllm/issues/6775 is a very good example of how to open an issue to ask for help in the right way. The clearer the issue descriptions, the faster you can get help.

There are many levels of environment variables:

1. environment variables set at machine start time (or container start time)
2. environment variables set by shell commands
3. environment variables set during program execution

This becomes complicated when we create multiple processes, and even create processes in another node when using `ray`. In distributed serving with multiple GPUs and multiple processes, **we need to make sure environment variables are set correctly for every process. Otherwise it might cause error**.

When vLLM starts a new process, what environment variables will the new process have? The complete answer is:

- multiprocessing backend: all 1/2/3 are inherited by the new process
- ray backend, with a local Ray instance: inherit 1/2/3 until `ray.init()`
- ray backend, with a pre-launched cluster: inherit environment variables when `ray start` is executed

Here is a short script to help people understand:

```python
# test.py

import os
import ray

def report_env(prefix="ray actor process"):
    keys = [
        "CONTAINER_ENV",
        "CMD_ENV",
        "DYNAMIC_ENV",
    ]
    for k in keys:
        v = os.environ.get(k, None)
        print(f"{prefix}: {k}={v}")

@ray.remote(num_gpus=1)
class Actor:
    def __init__(self) -> None:
        report_env()

def driver_function(value="DRIVER"):
    os.environ["DYNAMIC_ENV"] = value

if __name__ == "__main__":
    driver_function(value="Before ray")
    ray.init()
    driver_function(value="After ray")

    report_env(prefix="parent process")

    actors = [Actor.remote(), Actor.remote()]

    import multiprocessing

    context = multiprocessing.get_context("spawn")
    p = context.Process(target=report_env, args=("child process:",))
    p.start()
    p.join()
```

And here is the output of executing with and without ray cluster:

```shell
$ CMD_ENV=1 python test.py

2024-07-25 13:05:57,185 INFO worker.py:1788 -- Started a local Ray instance.
parent process: CONTAINER_ENV=None
parent process: CMD_ENV=1
parent process: DYNAMIC_ENV=After ray
child process:: CONTAINER_ENV=None
child process:: CMD_ENV=1
child process:: DYNAMIC_ENV=After ray
(Actor pid=4044013) ray actor process: CONTAINER_ENV=None
(Actor pid=4044013) ray actor process: CMD_ENV=1
(Actor pid=4044013) ray actor process: DYNAMIC_ENV=Before ray
(Actor pid=4044016) ray actor process: CONTAINER_ENV=None
(Actor pid=4044016) ray actor process: CMD_ENV=1
(Actor pid=4044016) ray actor process: DYNAMIC_ENV=Before ray

$ CMD_ENV=ray_start ray start --head

... output omitted ...

$ CMD_ENV=1 python test.py

2024-07-25 13:07:05,128 INFO worker.py:1603 -- Connecting to existing Ray cluster at address: xxx.xxx.xxx.xxx:6379...
2024-07-25 13:07:05,134 INFO worker.py:1788 -- Connected to Ray cluster.
parent process: CONTAINER_ENV=None
parent process: CMD_ENV=1
parent process: DYNAMIC_ENV=After ray
child process:: CONTAINER_ENV=None
child process:: CMD_ENV=1
child process:: DYNAMIC_ENV=After ray
(Actor pid=4058591) ray actor process: CONTAINER_ENV=None
(Actor pid=4058591) ray actor process: CMD_ENV=ray_start
(Actor pid=4058591) ray actor process: DYNAMIC_ENV=None
(Actor pid=4058593) ray actor process: CONTAINER_ENV=None
(Actor pid=4058593) ray actor process: CMD_ENV=ray_start
(Actor pid=4058593) ray actor process: DYNAMIC_ENV=None
```

As we can see, child process always inherits all environment variables from the parent process. The ray actor process, instead, only inherits the state when we create the cluster:

- when we start a local instance via `ray.init`, the cluster is started from the Python interpreter, and inherit the state at that moment. So we can see `ray actor process: DYNAMIC_ENV=Before ray` and `child process:: DYNAMIC_ENV=After ray`.
- when we start a ray cluster before we execute the python script, the ray actor process later also remember the cluster's state, but is not affected by the python process. So we can see `child process:: CMD_ENV=1` and `ray actor process: CMD_ENV=ray_start`.

The common usecase for single-node ray backend is to start a local instance via `ray.init`. In this regard, shell environment variables can be successfully applied to all vLLM processes.

> TL;DR: normally, in single-node case, setting environment variables before executing vLLM should just work as expected.

Now let's go to the multi-node case. We start a cluster by (the script and usage are described in the [doc](https://docs.vllm.ai/en/stable/serving/distributed_serving.html#multi-node-inference-and-serving)):

```shell
# head node
$ bash run_cluster.sh vllm/vllm-openai 10.150.0.34 --head ~/.cache/huggingface -v /home/gcpuser/vllm/examples:/root/cwd -e CONTAINER_ENV=head

# worker node
$ bash run_cluster.sh vllm/vllm-openai 10.150.0.34 --worker ~/.cache/huggingface -v /home/gcpuser/vllm/examples:/root/cwd -e CONTAINER_ENV=worker
```

Note that we defined different value for `CONTAINER_ENV` for different nodes. Let's see what would happen:

```shell
# head node
$ docker exec -it node /bin/bash

$ CMD_ENV=1 python /root/cwd/test.py

2024-07-25 21:29:24,521	INFO worker.py:1603 -- Connecting to existing Ray cluster at address: 10.150.0.34:6379...
2024-07-25 21:29:24,529	INFO worker.py:1788 -- Connected to Ray cluster.
parent process: CONTAINER_ENV=head
parent process: CMD_ENV=1
parent process: DYNAMIC_ENV=After ray
child process:: CONTAINER_ENV=head
child process:: CMD_ENV=1
child process:: DYNAMIC_ENV=After ray
(Actor pid=100, ip=10.150.0.35) ray actor process: CONTAINER_ENV=worker
(Actor pid=100, ip=10.150.0.35) ray actor process: CMD_ENV=None
(Actor pid=100, ip=10.150.0.35) ray actor process: DYNAMIC_ENV=None
(Actor pid=314) ray actor process: CONTAINER_ENV=head
(Actor pid=314) ray actor process: CMD_ENV=None
(Actor pid=314) ray actor process: DYNAMIC_ENV=None
```

Given the result, all processes in the head node get `CONTAINER_ENV=head`, while the process in the worker node gets `CONTAINER_ENV=worker`. Meanwhile, all the Python processes get `CMD_ENV=1`, but all the ray actor processes get `CMD_ENV=None`.

Therefore, the conclusion is: if we start a multi-node cluster, the only reliable way of setting environment variables , is to set environment variables at the container start time. All the processes in the node, will use the environment variables set by that node.

This is important for dealing with complicated network configurations. For example, if the head node uses `eth0` network interface to talk to `eth1` network interface in the worker node, then we should launch the cluster as:

```shell
# head node
$ bash run_cluster.sh vllm/vllm-openai 10.150.0.34 --head ~/.cache/huggingface -v /home/gcpuser/vllm/examples:/root/cwd -e NCCL_SOCKET_IFNAME=eth0 -e GLOO_SOCKET_IFNAME=eth0

# worker node
$ bash run_cluster.sh vllm/vllm-openai 10.150.0.34 --worker ~/.cache/huggingface -v /home/gcpuser/vllm/examples:/root/cwd -e NCCL_SOCKET_IFNAME=eth1 -e GLOO_SOCKET_IFNAME=eth1
```

> TL;DR: in multi-node case, we need to set environment variables when we start the container. Processes in each container will use the value of environment variables defined by that container.
